### ISSUE TITLE
Testing high-order procedure for gradient calculation

### ISSUE THREAD

**Issue by iBatistic** [philipcardiff/paper-newton-krylov#9]
Here are some initial results; I’ll get back to you soon after I check in more detail how it works. I haven’t yet added treatment for boundary cells and faces on the edge. If you're interested in the code, you can check it on my `solids4foam-ib` fork on the branch `solids4foam/feature-petsc-snes` in `linGeomTotalDispSolid.C` from line 220 to 475 (everything is still in one place).

<img src="https://github.com/user-attachments/assets/0104a9a7-22c3-4bc7-844a-04f19f92116f" width="600">

**Comment by philipcardiff** [comment]
This looks really impressive.

The code is nicely structured and easy to follow.

By the way, I was recently reading the [Zienkiewicz book (volume 1)](https://www.sciencedirect.com/book/9781856176330/the-finite-element-method-its-basis-and-fundamentals) and interesting they suggest a similar least squares approach for "recovering" the stress from the displacement field (in [Chapter 15 - Errors, Recovery Processes, and Error Estimates](https://www.sciencedirect.com/science/article/pii/B9781856176330000150)). Essentially, they can get more accurate (and higher order of accuracy) stresses by fitting a least squares plane about each vertex through the surrounding elements' Gauss points, and this is how they interpolate stresses from Gauss points to the mesh vertices.

**Comment by iBatistic** [comment]
Hi Philip,

I've been occupied with other tasks lately, so I haven't been able to finish this yet. This morning, I added the treatment for boundary cells, and the results seem to be coming out well. I still need to incorporate the treatment of boundary faces to compute the boundary face gradient.

> By the way, I was recently reading the [Zienkiewicz book (volume 1)](https://www.sciencedirect.com/book/9781856176330/the-finite-element-method-its-basis-and-fundamentals) and interesting they suggest a similar least squares approach for "recovering" the stress from the displacement field (in [Chapter 15 - Errors, Recovery Processes, and Error Estimates](https://www.sciencedirect.com/science/article/pii/B9781856176330000150)).

Thank you for this. Yep, it looks like this!

This is a comparison between s4f `leastSquares` and HO interpolation method with 2 layers of cells and a linear variable distribution.
<img src="https://github.com/user-attachments/assets/1421a473-0f06-43cb-9d66-ecd2f5d63aa4" width="800">

This is comparison with HO interpolation with 4 layers of cells and a quadratic variable distribution.
<img src="https://github.com/user-attachments/assets/77b02295-0b7b-437c-8cd5-6c40a5d8aebd" width="800">

**Comment by philipcardiff** [comment]
No problem.

This looks great. It is hard to see but it looks like the HO method gives better predictions near the poor cells (where the blockMesh blocks meet). For the boundary faces, I suggest that initially, we just use the adjacent cell-centre gradient; in fact, this is what we are doing in the current NK method. It will be interesting to see how the errors look on the MMS case with the unstructured mesh.

By the way, I am slowly making progress on the Newton-Krylov paper and hope to have a full draft in the coming weeks, as my teaching comes to an end.

**Comment by iBatistic** [comment]
I just realized I didn’t reply on Friday—I must have forgotten to hit the comment button below! 😅

> For the boundary faces, I suggest that initially, we just use the adjacent cell-centre gradient; in fact, this is what we are doing in the current NK method

Then we can go ahead and test it. I ran it on a 3D case, and it worked well. The code is still in one place in my solids4foam fork (branch feature-petsc-snes), so we can place it into a single function and call it as needed.

Are you planning to try this method for real-time gradient calculation, or do you see it more as a postprocessing step for stress distribution?

**Comment by philipcardiff** [comment]
OK, great.

It will be interesting to try it out as both a post-processing step and part of the solution. If it does not work as part of the solution, then at least it will be useful for post-processing.

I will look to merge it into the petsc branch and also think about the parallelization. I will keep you posted.

**Comment by iBatistic** [comment]
Ok, let me know what results you get.

p.s The only parameters controlling the interpolation results are:
```
//Number of layers in interpolation molecule
 const label nLayers = 4;
//Order of interpolation (1 is for linear variable distribution and 2 is for quadratic)
const label N = 2;
```

**Comment by philipcardiff** [comment]
I wrapped your code up into a class: see here: https://github.com/solids4foam/solids4foam/blob/feature-petsc-snes-higher-order/src/solids4FoamModels/solidModels/linGeomTotalDispSolid/higherOrderGrad/higherOrderGrad.H

You will then see in the `residualMomentum` function in `linGeomTotalDispSolid` that I call your higher order gradient instead of the normal gradient procedure, e.g.:
```
    // Update gradient of displacement
    //mechanical().grad(D, gradD());
    gradD() = hoGrad_.grad(D);
```

It runs and converges (which is really promising), but it seems to converge to a different answer for the MMS case. I also tried with `N = 1` and `nLayers = 1` (so it should be basically the same as our current least squares), but it also converges to the same different answer. If you have any thoughts, let me know before I look deeper into what's happening.

**Comment by iBatistic** [comment]
Nice, ! I’ll take a look today and try to figure out why that’s happening

**Comment by iBatistic** [comment]
Hi @philipcardiff ,

this line:
`gradD[cellI] += D[curStencil[cI]]*interpGradCoeffs[cellI][cI];`
and this line:
`gradD[cellI] +=boundaryD*interpGradCoeffs[cellI][cI];`

are wrong. They should be like this:

`gradD[cellI] += *interpGradCoeffs[cellI][cI]*D[curStencil[cI]];`
and:
`gradD[cellI]+=interpGradCoeffs[cellI][cI]*boundaryD;`

With this correction, on MMS case I'm getting this convergence for displacement (4 layers, 2 order):
![Screenshot from 2024-11-13 11-55-50](https://github.com/user-attachments/assets/acbb8984-4618-45e8-8635-8841461ed239)


**EDIT:** The above diagram is generated using a tet mesh. With a hex mesh, I got slightly bigger errors but the slope is the same. tet is better for interpolation because the molecule is not structured resulting in lower conditional number.

**Comment by philipcardiff** [comment]
Great; I will play with this change later and let you know how it goes.

**Comment by iBatistic** [comment]
I noticed yesterday that other cases are not working. When you confirm that you get good results on the MMS case we need to check what's going on with the other. I guess it's something with the traction boundaries as the MMS case does not have it.

**Comment by philipcardiff** [comment]
In the current leastSquaresS4f, only fixedValue boundary values are included in the least squares stencil; traction and other boundary face valuea are not, since we do not know the value there (and instead we will extrapolate using the gradient). I suspect this is the cause, as MMS does not have traction boundaries, as you said.

**Comment by philipcardiff** [comment]
I pushed a change here: https://github.com/solids4foam/solids4foam/tree/feature-petsc-snes-higher-order/src/solids4FoamModels/solidModels/linGeomTotalDispSolid/higherOrderGrad

where the patches to be included in the stencil are indicated by the `includePatchInStencils` boolList. Also, I made the hoGrad object a pointer so it can be optionally enabled from the case.

The method seems to work, which is cool.
By the way, I noticed that the calcCoeffs function becomes slow when four layers are added (but the make stencils is not that slow); I wonder what the slow part is; I suspect it is related to the DynamicLists.

**Comment by philipcardiff** [comment]
From some quick timing measurements (i.e., `mesh.time().clockTimeIncrement();`), I found that the following line seems to be the reason for the slow down:
```
	t2 += mesh.time().clockTimeIncrement();
        const Eigen::MatrixXd Bhat =
            sqrtW*Eigen::MatrixXd::Identity(W.rows(), W.cols());
        t3 += mesh.time().clockTimeIncrement();
```
I got this timings for the MMS case mesh 3 with 4 layers and N = 2:
```
t1 = 0
t2 = 3
t3 = 9
t4 = 1
t5 = 0
```
which basically says most of the time (9 s) occurred between t2 and t3.

It seems that it creates a 368 x 368 matrix for some cells, which is quite huge. I have not gone through the algorithm properly yet so this may be unavoidable.

**Comment by philipcardiff** [comment]
ChatGPT helped:
```
        // const Eigen::MatrixXd Bhat =                                                                                   
        //     sqrtW*Eigen::MatrixXd::Identity(W.rows(), W.cols());                                                       
        const Eigen::MatrixXd Bhat = sqrtW.diagonal().asDiagonal();
```

Now the timings are:
```
t1 = 0
t2 = 2
t3 = 0
t4 = 2
t5 = 0
```

Now to see if ChatGPT can help with t2 and t4!

Edit: FYI, the total run time is now 5 seconds for this case so you can see 4 s (2 + 2) still comes from calcCoeffs!

**Comment by iBatistic** [comment]
Oops, I didn't think the line with Bhat could slow things down that much.
 Do these 4 seconds also come from lines where Eigen is used?

**Comment by philipcardiff** [comment]
OK, I made a few more changes with the help of ChatGPT, and it is a bit faster; using diagonal matrices, possibly, helps a lot. But the `Rbar.colPivHouseholderQr` is still expensive. For example, 50% of the time of the entire solver is spent in this function.

**Comment by iBatistic** [comment]
Castillo mentioned in the paper that QR is expensive. I guess one solution for speed-up can be defining a threshold above which QR is used. 

![Screenshot from 2024-11-14 23-09-36](https://github.com/user-attachments/assets/d320ff9c-8822-4e25-9cd7-b57b7a925155)

Did you try 1 layer and 1 order? If I remember correctly, I got a bit lower error than leastSquares which is interesting.

**Comment by philipcardiff** [comment]
Interesting. That's good to know.  One option is that we implement a flag to allow us to switch between QR and Cholesky; alternatively, we would have a smart flag that automatically switches to QR, although I am not sure how we would know to switch (without doing a QR decomposition or similar).

There are other things we can do too if this becomes an issue, e.g., store the coefficients on disk and read them if present for subsequent cases, and possibly create a utility for precomputing the coefficients (this could even use OpenMP parallelisation as this task should parallelise perfectly, assuming no domain decomposition).

I wonder what other authors do, e.g. higher order CFD FVMs.

I did not realise that 1 layer and order 1 is more accurate than our default method, interesting. I guess the weights may be better; also, this new method fits four parameters for order 1 (offset and three slopes), whereas OpenFOAM assumes the offset and only calculates the slopes.

**Comment by philipcardiff** [comment]
> Oops, I didn't think the line with Bhat could slow things down that much. Do these 4 seconds also come from lines where Eigen is used?

I realised that the `clockTimeIncrement()` approach was not ideal for this task due to rounding of the numbers, so I switched to a Time Profiler app that is built-in on Mac. This gave me the exact breakdown time for each function and, yep, the Eigen QR decomposition is the most expensive part by far. I depends on how fast or slow the subsequent solve step is to determine whether it is a bottleneck or not. For transient cases it is likely not an issue because the coefficients are only calculated once, but you can definitely notice the calcCoeffs time for the MMS cases when nLayers = 4 and N = 2.

**Comment by philipcardiff** [comment]
With lots of help from ChatGPT, I was able to add Cholesky decomposition as an option, which is enabled by setting the new `useQRDecomposition_` flag to "off".

For the MMS cases, it seems to produce exactly the same output (max stress error is the same to the number of figures printed to the screen). You can notice it is faster as nLayers gets bigger (13 vs 23 s for N = 2, nLayers = 5), but the speedup is small for smaller nLayer and N.

Interestingly, ChatGPT gave some nice suggestions for efficiently checking when we should switch from Cholesky to QR, e.g. perform Cholesky first and switch to QR if it fails, or add regularisation to the Cholesky N matrix, e.g. `N += lambda * Eigen::MatrixXd::Identity(N.rows(), N.cols());`.

It looks like we have lots of options to play with.

Edit: By the way, you will see that the Cholesky decomposition requires the storage of different matrices, so it might be cleaner if we separate the Cholesky and QR codes into different functions.

**Comment by iBatistic** [comment]
@philipcardiff, there’s also a third option. After obtaining the 𝑊 and 𝑄 matrices, we can calculate the 𝑀 matrix as follows:
```
M = Q @ W @ Q.T 
 ```

Then, the 𝐴 matrix can be computed like this:
```
A = np.linalg.pinv(M) @ Q @ W  
```
In Python, I checked the condition number of 𝑀, and if it was below 1e5, I used `pinv` instead of `inv`. This approach results in a larger condition number, but I guess it should be faster. For unstructured meshes, this might be a suitable approach.
I started exploring the QR method after noticing the high condition number on structured meshes. However, my stencil construction was based on a sphere approach rather than a layer approach (which is better). We should also try this to see what results we get.

**Comment by philipcardiff** [comment]
Interesting.

I think "np.linalg.pinv(M)" uses SVD, which I guess will be similar in speed to (or slower than) QR decomposition, but I think SVD is the most stable as it will even give reasonable answers for rank deficient matrices (e.g. not enough points in the stencil).

Given our planned research directions, it will be worthwhile comparing each of these approaches in terms of robustness, accuracy and efficiency.

... I am also starting to think about parallelisation ...

**Comment by iBatistic** [comment]
> Edit: By the way, you will see that the Cholesky decomposition requires the storage of different matrices, so it might be cleaner if we separate the Cholesky and QR codes into different functions.

Sure, I agree

> There are other things we can do too if this becomes an issue, e.g., store the coefficients on disk and read them if present for subsequent cases, and possibly create a utility for precomputing the coefficients (this could even use OpenMP parallelisation as this task should parallelise perfectly, assuming no domain decomposition).
> I wonder what other authors do, e.g. higher order CFD FVMs.

Hmm, yes, that’s a neat idea—precomputing the coefficients and storing them. The calculation time for these coefficients is likely not an issue with a static mesh and a large number of iterations.

What do you suggest we do next? Should I try to fix the implementation to work with `solidTraction` boundaries?

Other than that, we could check what results we get with third-order accuracy. For this, we just need to include the Taylor coefficients (the remaining code should be the same). We could experiment with this approach, as it’s not necessary to include the entire row of the Pascal triangle.

I also noticed that some authors incorporate multiple boundary face values (from the first layer of cells) into the boundary cell stencil. Intuitively, I believe this might help reduce the 𝐿∞  error in the MMS case.

**Comment by iBatistic** [comment]
> I think "np.linalg.pinv(M)" uses SVD, which I guess will be similar in speed to (or slower than) QR decomposition, but I think SVD is the most stable as it will even give reasonable answers for rank deficient matrices (e.g. not enough points in the stencil).

Ah yes, I forgot that.

**Comment by philipcardiff** [comment]
> > There are other things we can do too if this becomes an issue, e.g., store the coefficients on disk and read them if present for subsequent cases, and possibly create a utility for precomputing the coefficients (this could even use OpenMP parallelisation as this task should parallelise perfectly, assuming no domain decomposition).
> > I wonder what other authors do, e.g. higher order CFD FVMs.
> 
> Hmm, yes, that’s a neat idea—precomputing the coefficients and storing them. The calculation time for these coefficients is likely not an issue with a static mesh and a large number of iterations.
> 
> What do you suggest we do next? Should I try to fix the implementation to work with `solidTraction` boundaries?

I included `includePatchInStencils`, which I think has solved the problem with solidTraction. Essentially, we just extrapolate at traction boundaries, which I think makes sense. Separately, when discretising the governing equation, we directly enforce the true traction at traction boundaries, which means it does not depend on the discretisation.

> 
> Other than that, we could check what results we get with third-order accuracy. For this, we just need to include the Taylor coefficients (the remaining code should be the same). We could experiment with this approach, as it’s not necessary to include the entire row of the Pascal triangle.

That sounds like a nice idea; are you happy to try this? And I will think about parallelisation: I will see if ChatGPT can advise on how best to collect and store off-proc data. 

> I also noticed that some authors incorporate multiple boundary face values (from the first layer of cells) into the boundary cell stencil. Intuitively, I believe this might help reduce the 𝐿∞ error in the MMS case.

Interesting. I think you are right, that additional fixedValue boundary faces will increase the accuracy (although the order of accuracy will probably not change).

**Comment by philipcardiff** [comment]
I ask ChatGPT about generalising the generation of the Taylor series terms and assignment of the Q matrix entries and it gave a straight-forward solution. So I tried it out and it worked (with a couple of iterations): see the latest commit. One thing to note is that the quadratic and higher order terms may be in a slightly different order than you manually set them, e.g. XY, YZ, XZ, etc., but this should not matter since we are only using the first gradients. Even if we do later want the second gradients, we can figure out the order then.

From a quick check on the MMS case with a polyhedral grid, it is possible to get smaller stress erorrs using N = 3 or 4 than N = 2, which is cool. Interestingly it seems that increasing the numbers of layers reduces the max stress error but increases the average stress errors; I guess this makes sense as the max stress errors are likely on the boundary, where the stencil is smaller. This suggests that we could consider automatically setting `nLayers` to be 1 higher for boundary cells.

I have not checked the order of accuracy for N > 2 but I assume the behaviour is still second order, since we are not using quadrature (for tractions of the MMS body force term). Another nice thing is that the JFNK procedure still seems to converge fine. I guess implementing an explicit Gauss quadratic scheme at the faces would give us true higher order :D

ChatGPT suggested a nice approach for parallelising the stencil generation so I will get to that next. It is amazing how much it can affect coding (and other) productivity.

Now to sleep!

**Comment by iBatistic** [comment]
> I included includePatchInStencils, which I think has solved the problem with solidTraction. Essentially, we just extrapolate at traction boundaries, which I think makes sense. Separately, when discretising the governing equation, we directly enforce the true traction at traction boundaries, which means it does not depend on the discretisation.

Oh yes, that's the correct solution. I completely forgot that `solidTraction` patches shouldn’t be included.
`

> I ask ChatGPT about generalising the generation of the Taylor series terms and assignment of the Q matrix entries and it gave a straight-forward solution. So I tried it out and it worked (with a couple of iterations): see the latest commit. One thing to note is that the quadratic and higher order terms may be in a slightly different order than you manually set them, e.g. XY, YZ, XZ, etc., but this should not matter since we are only using the first gradients. Even if we do later want the second gradients, we can figure out the order then.

You rearranged the code very nicely! As you said, the order of the members doesn’t matter; we just need to know where the required component is, and that’s it.

> From a quick check on the MMS case with a polyhedral grid, it is possible to get smaller stress erorrs using N = 3 or 4 than N = 2, which is cool. Interestingly it seems that increasing the numbers of layers reduces the max stress error but increases the average stress errors; I guess this makes sense as the max stress errors are likely on the boundary, where the stencil is smaller. This suggests that we could consider automatically setting nLayers to be 1 higher for boundary cells.

Apart from this, we should definitely try including more boundary face values to see the results we get. Ultimately, it’s up to us to decide on these things, as they are not strictly defined mathematically

> I have not checked the order of accuracy for N > 2 but I assume the behaviour is still second order, since we are not using quadrature (for tractions of the MMS body force term). Another nice thing is that the JFNK procedure still seems to converge fine. I guess implementing an explicit Gauss quadratic scheme at the faces would give us true higher order :D

For the cubic and quartic Taylor expression, the order of convergence for the displacement is approximately 2.25 - 2.3, whereas the stress remains at 1.5. It makes sense to me that if the solution variable is second-order, its derivative is unlikely to achieve the same order. However, you mentioned that Zeljko was obtaining second-order convergence for both stress and displacements, which I don’t fully understand.

> . I guess implementing an explicit Gauss quadratic scheme at the faces would give us true higher order :D

Then the gradient at the cell center would be calculated using Gauss's theorem?

**Comment by philipcardiff** [comment]
> > I have not checked the order of accuracy for N > 2 but I assume the behaviour is still second order, since we are not using quadrature (for tractions of the MMS body force term). Another nice thing is that the JFNK procedure still seems to converge fine. I guess implementing an explicit Gauss quadratic scheme at the faces would give us true higher order :D
> 
> For the cubic and quartic Taylor expression, the order of convergence for the displacement is approximately 2.25 - 2.3, whereas the stress remains at 1.5. It makes sense to me that if the solution variable is second-order, its derivative is unlikely to achieve the same order. However, you mentioned that Zeljko was obtaining second-order convergence for both stress and displacements, which I don’t fully understand.

I asked Zeljko about this and he confirmed that he only received 2nd order accuracy for the boundary stresses at the hole in the plateHole and he did not examined the stress errors in the internal cells, e.g. where the cells are skewed. I was able to run the extend-bazaar code and case and it shows that the max stress errors are closer to 1st order, as expected. So our results are consistent with his findings.


> > . I guess implementing an explicit Gauss quadratic scheme at the faces would give us true higher order :D
> 
> Then the gradient at the cell center would be calculated using Gauss's theorem?

I was thinking about calculating the gradient at the faces using the higher-order approach. I suppose in that case we would build the least squares method directly at the face (as you did in the Python code). We would then need quadrature at the faces, like you did. However, based on our recent experiences, I suspect that we would only need the explicit implementation as the JFNK method will probably work. For the cell-centred gradients (post processing), I think our current approach is fine. I guess the Gauss's theorem method would also work, albeit it would be more expensive, and I guess it would instead give the average gradient in each cell rather than the gradient at the cell-centre (both are fine but represent different things).

Generalising the least squares method to the faces seems like an interesting next step ...

**Comment by iBatistic** [comment]
> I asked Zeljko about this and he confirmed that he only received 2nd order accuracy for the boundary stresses at the hole in the plateHole and he did not examined the stress errors in the internal cells, e.g. where the cells are skewed. I was able to run the extend-bazaar code and case and it shows that the max stress errors are closer to 1st order, as expected. So our results are consistent with his findings.

Ah, okay then. I didn't know.

> I was thinking about calculating the gradient at the faces using the higher-order approach. I suppose in that case we would build the least squares method directly at the face (as you did in the Python code). We would then need quadrature at the faces, like you did. However, based on our recent experiences, I suspect that we would only need the explicit implementation as the JFNK method will probably work.

Yes, we will create a face molecule that remains the same for all face Gauss points, and we will calculate interpolation coefficients for each Gauss point. With the coefficients for each Gauss point, we can perform integration in an implicit manner (using a block solver - like in Python code) or integrate explicitly, as you mentioned. Explicit integration sounds interesting; this approach seems like it can open a bunch of possibilities.

I noticed you were working on parallelisation; nice!

For fixed displacement boundaries, I tried to add boundary faces to the interpolation molecule of the boundary cells, using only the faces from the first layer of the interpolation molecule. However, the improvement in results was not as expected. While the displacement error is slightly reduced, the stress error is slightly higher. I suspect that the increase in the condition number is the cause.

**Comment by philipcardiff** [comment]
> > I was thinking about calculating the gradient at the faces using the higher-order approach. I suppose in that case we would build the least squares method directly at the face (as you did in the Python code). We would then need quadrature at the faces, like you did. However, based on our recent experiences, I suspect that we would only need the explicit implementation as the JFNK method will probably work.
> 
> Yes, we will create a face molecule that remains the same for all face Gauss points, and we will calculate interpolation coefficients for each Gauss point. With the coefficients for each Gauss point, we can perform integration in an implicit manner (using a block solver - like in Python code) or integrate explicitly, as you mentioned. Explicit integration sounds interesting; this approach seems like it can open a bunch of possibilities.

OK, great. I think we can re-use the cell stencils for the face stencils, e.g. we can combine the stencils of the two cells sharing a face to get the face stencil. This way, the face stencil function should be straight-forward (I think...).


> I noticed you were working on parallelisation; nice!

Thanks. I pushed a code to use the global cell stencils to calculate the gradient using the Cholesky approach (e.g., the `gradGlobalCholeksy` function); it seems to work, but I have not fully tested it. You will see that `globalCellStencils` gives the stencils in terms of global cell IDs; when we use them, we need to check if they are on the current processor (`globalCells_.isLocal(globalCellID)`) and, if needed, convert them to a local cell ID (`globalCells_.toLocal(globalCellID)`). For cells which are not on this processor, I created a function to get field data from other processors. This all seems to work, but I am not sure the parallel communication and data storage approaches are the most efficient choice, i.e., I am using `Map`, which is a `HashTable`, and ideally, I want to perform proc-to-proc communication rather than global scatter-gather. But it's a start, and ChatGPT will help me improve it if needed.
 
> For fixed displacement boundaries, I tried to add boundary faces to the interpolation molecule of the boundary cells, using only the faces from the first layer of the interpolation molecule. However, the improvement in results was not as expected. While the displacement error is slightly reduced, the stress error is slightly higher. I suspect that the increase in the condition number is the cause.

Interesting.

Including the boundary face values is a bit of a pain from a coding perspective (and finding them in parallel, too). It may be sufficient (from an order of accuracy perspective) to just use the cell values. But we can also examine this more and check what others have done.

It looks like we have made a lot of progress recently with this; what approach do you propose for implementing explicit Gauss quadrature at the faces? I guess we need to decompose the faces on-the-fly into shapes that we can apply Gauss quadrature to, and then calculate the gradients at these points, and then feed these to the material model (this will add further motivation to generalising the material model so we can pass any Field size as input and get back a stress Field of the same size).

I am still finishing work on the paper; possibly, this higher-order approach could be incorporated, which would be a major contribution, or it may be better to keep it for a second paper.

**Comment by iBatistic** [comment]
> OK, great. I think we can re-use the cell stencils for the face stencils, e.g. we can combine the stencils of the two cells sharing a face to get the face stencil. This way, the face stencil function should be straight-forward (I think...).

Nice idea! I can't think of any reason at the moment not to do it that way.

> Thanks. I pushed a code to use the global cell stencils to calculate the gradient using the Cholesky approach (e.g., the gradGlobalCholeksy function); it seems to work, but I have not fully tested it. You will see that globalCellStencils gives the stencils in terms of global cell IDs; when we use them, we need to check if they are on the current processor (globalCells_.isLocal(globalCellID)) and, if needed, convert them to a local cell ID (globalCells_.toLocal(globalCellID)). For cells which are not on this processor, I created a function to get field data from other processors. This all seems to work, but I am not sure the parallel communication and data storage approaches are the most efficient choice, i.e., I am using Map, which is a HashTable, and ideally, I want to perform proc-to-proc communication rather than global scatter-gather. But it's a start, and ChatGPT will help me improve it if needed.

Thanks for the explanation. It will make it easier for me to understand it.

At the end, should we move this code as a gradient calculation method and call it in `fvSchemes` dict?

> I am still finishing work on the paper; possibly, this higher-order approach could be incorporated, which would be a major contribution, or it may be better to keep it for a second paper.

Since most of the text is already finished and doesn't address this topic, I think incorporating this into the article might be "difficult" as "additional text" will differ from the current content. What do you think, would this gradient calculation method be enough on its own for a new article?

Have you tried it on the turbine blade case? Did it help with stress contours?



> It looks like we have made a lot of progress recently with this; what approach do you propose for implementing explicit Gauss quadrature at the faces? I guess we need to decompose the faces on-the-fly into shapes that we can apply Gauss quadrature to, and then calculate the gradients at these points, and then feed these to the material model (this will add further motivation to generalising the material model so we can pass any Field size as input and get back a stress Field of the same size).

In the Python code, I was decomposing faces using central point decomposition. For each triangle, we can decide how many Gauss points to use. The overall face integral is then the sum of the integrals over all the triangles. For example, a quadrilateral face can be decomposed into four triangles, with each triangle having, let's say, five Gauss points. In total, that would be 20 Gauss points per face.
This will increase memory requirements, as each Gauss point needs its own weights. I can create the first version of the code for this approach, but you need to tell me where. I guess that the volume field should generate and store the interpolation coefficients for each Gauss point, while the integration itself should be handled separately somewhere in the code.

**Comment by philipcardiff** [comment]
> At the end, should we move this code as a gradient calculation method and call it in `fvSchemes` dict?

Good question. That would probably make sense. You will see that I have now added many functions (with `if-else` checks), so the class is becoming a bit long and unwieldy. Refactoring into a derived class structure will likely make it easier to follow and understand.
However, before doing this, let's get the face gradient method working, as this may affect our design choices.

 
> > I am still finishing the paper; this higher-order approach could be incorporated, which would be a major contribution, or it may be better to keep it for a second paper.
> 
> Since most of the text is already finished and doesn't address this topic, I think incorporating this into the article might be "difficult" as "additional text" will differ from the current content. What do you think, would this gradient calculation method be enough on its own for a new article?

Good point. So far, I only tried it on the MMS case, but, I agree, it would be better to finish the paper since it is close. We can then leave generalised higher-order work for another article, maybe also looking at JFNK. I have noticed that the convergence (of the linear solver) in the axial turbine is worse than other cases, and suspect it may be related to mesh non-orthogonality, and hence the preconditioner implicit Laplacian matrix is not as good. Including non-orthogonality corrections in this matrix may help a lot, but let's leave this to future work!


> Have you tried it on the turbine blade case? Did it help with stress contours?

Not yet. I will. I also need to think about which times/memory to report in the paper, if plan to present two gradient schemes (linear and quadratic LS).


> In the Python code, I was decomposing faces using central point decomposition. For each triangle, we can decide how many Gauss points to use. The overall face integral is then the sum of the integrals over all the triangles. For example, a quadrilateral face can be decomposed into four triangles, with each triangle having, let's say, five Gauss points. In total, that would be 20 Gauss points per face. This will increase memory requirements, as each Gauss point needs its own weights. I can create the first version of the code for this approach, but you need to tell me where. I guess that the volume field should generate and store the interpolation coefficients for each Gauss point, while the integration itself should be handled separately somewhere in the code.

OK, that makes sense to me.

You will see that I pushed an initial face gradient function `tmp<surfaceTensorField> higherOrderGrad::fGrad(const volVectorField& D);`, where only one Gauss point is assumed, i.e. gradient at the face centroid. I am thinking of the following design (let me know your thoughts):
- Create a similar function called `fGradGaussPoints`, which calculates gradD at the face Gauss points and returns them as a Field, e.g.  
  - `tmp<Field<tensor>> higherOrderGrad::fGradGaussPoints(const volVectorField& D);`
  -  where the returned field has the size `label size = 0; forAll(faceGaussPoints(), faceI){size += faceGaussPoints()[faceI].size();}` with `faceGaussPoints()` being a `List<pointField>` of the Gauss point coordinates per face (or a similar field, if we do not need to store the positions).
- This `fGradGaussPoints` function will need the coefficients at the face Gauss points. We can create and store these just like we currently do for the cells, however, in this case, we have an extra nested level.
- In the solid model (e.g. `linGeomTotalDispSolid`), we can call the`fGradGaussPoints` function to get gradD at the face Gauss points.
- We then need to add an interface to the mechanical model (something simple initially for testing) that takes a `Field<tensor>` as an input and returns the stress field as `Field<symmTensor>`, i.e., the stress at the face Gauss points.
- Finally, we need to define a function (e.g., `tmp<volVectorField> fvc::divGaussPoints(const Field<vector>& tractionAtGaussPoints)` or similar that lets us integrate the force at the faces. This function will need the Gauss weights, so will have to figure out the best place to calculate and store these (we can pass them as an argument if needed).

Does that make sense? Am I missing anything? Probably, we will learn as we start to implement it.

If you like, you can start on the `fGradGaussPoints` function and I will look at the material model and then the `divGaussPoints` function (we may need to discuss this to ensure consistency with the `fGradGaussPoints` approach).

By the way, for the paper authors, if you agree, I propose to move your name to last; that way, it will appear as a `senior` author publication for you, i.e. the first and last names are the most important. I know things like this can help later for promotions, etc. Of course, I don't mind either way.

**Comment by iBatistic** [comment]
> Good question. That would probably make sense. You will see that I have now added many functions (with if-else checks), so the class is becoming a bit long and unwieldy. Refactoring into a derived class structure will likely make it easier to follow and understand.
> However, before doing this, let's get the face gradient method working, as this may affect our design choices.

Sure, I agree.

> You will see that I pushed an initial face gradient function tmp<surfaceTensorField> higherOrderGrad::fGrad(const volVectorField& D);, where only one Gauss point is assumed, i.e. gradient at the face centroid. I am thinking of the following design (let me know your thoughts):.....

Currently it makes sense to me; I will have doubts when I will start but we can solve them on the fly.
I have just one question. Instead of `tmp<Field<tensor>>`, wouldn't it be better to use something like `tmp<Field<List<tensor>>`? In that way, each item in the field represents a face and each face stores values for Gauss points.

> If you like, you can start on the fGradGaussPoints function and I will look at the material model and then the divGaussPoints function (we may need to discuss this to ensure consistency with the fGradGaussPoints approach).

I will start working on the `fGradGaussPoints` function tomorrow :) I’ll keep you updated.

> By the way, for the paper authors, if you agree, I propose to move your name to last; that way, it will appear as a senior author publication for you, i.e. the first and last names are the most important. I know things like this can help later for promotions, etc. Of course, I don't mind either way.

Thank you for offering that; kind of you. From what I currently know, in our system, for each paper, we need to collect signatures from all authors to confirm who the two main authors are before the main author can use the paper for collecting promotion points. For papers with up to four authors, the full amount of points is awarded. However, if there are more than four authors, the points are proportionally reduced based on the total number of authors. Anyway, I’ll double-check if this is important in our system and let you know.

**Comment by philipcardiff** [comment]
> > You will see that I pushed an initial face gradient function tmp higherOrderGrad::fGrad(const volVectorField& D);, where only one Gauss point is assumed, i.e. gradient at the face centroid. I am thinking of the following design (let me know your thoughts):.....
> 
> Currently it makes sense to me; I will have doubts when I will start but we can solve them on the fly. I have just one question. Instead of `tmp<Field<tensor>>`, wouldn't it be better to use something like `tmp<Field<List<tensor>>`? In that way, each item in the field represents a face and each face stores values for Gauss points.

You are right; that makes more sense. We can store the Gauss weights and points (if needed) in a similar structure.

> Thank you for offering that; kind of you. From what I currently know, in our system, for each paper, we need to collect signatures from all authors to confirm who the two main authors are before the main author can use the paper for collecting promotion points. For papers with up to four authors, the full amount of points is awarded. However, if there are more than four authors, the points are proportionally reduced based on the total number of authors. Anyway, I’ll double-check if this is important in our system and let you know.

Ah, OK. Before submission, we can check that it aligns with your internal policies.

**Comment by iBatistic** [comment]
@philipcardiff, how would it be best to store the Gaussian integration points' locations (in natural coordinates) and their corresponding weighting factors?

Here is Python implementation:
```
# Dunavant quadrature points and weights for different orders of precision
TRIANGLE = \
{
    1:
        {
            'points': np.array([[0.333333333333333, 0.333333333333333, 0.333333333333333]]),
            'weights': np.array([1])
        },
    2:
        {
            'points': np.array([
                [0.666666666666667, 0.166666666666667, 0.166666666666667],
                [0.166666666666667, 0.666666666666667, 0.166666666666667],
                [0.166666666666667, 0.166666666666667, 0.666666666666667]
            ]),
            'weights': np.array([
                0.333333333333333,
                0.333333333333333,
                0.333333333333333
            ])
        },
    3:
        {
            'points': np.array([
                [0.333333333333333, 0.333333333333333, 0.333333333333333],
                [0.600000000000000, 0.200000000000000, 0.200000000000000],
                [0.200000000000000, 0.600000000000000, 0.200000000000000],
                [0.200000000000000, 0.200000000000000, 0.600000000000000]
            ]),
            'weights': np.array([
                -0.5625,
                0.520833333333333,
                0.520833333333333,
                0.520833333333333
            ])
        }
}

```

 I’ve been thinking for a while about the appropriate way to handle this; I’ve created a `triangleQuadrature` class to handle functionalities related to mapping between natural and cartesian coordinates. My idea was to store the natural coordinates of Gauss points as a static variable. Is this something you would approach in the same way, or do you have a different approach in mind?


**UPDATE**
Philip, you can ignore the question above, as I've [solved](https://github.com/iBatistic/solids4foam-ib/commit/8012962a9268aa06f1ffb67efeec14ebf13ccd36) it for now. When I finish you can refactor the implementation. For now, I've implemented a solution using a Map object that stores a structure containing integration points and weight factors.

**Comment by iBatistic** [comment]
Just to refresh my memory of where we left off; we have implemented the following points:

> - [x] Create a similar function called `fGradGaussPoints`, which calculates gradD at the face Gauss points and returns them as a Field, e.g.  
> - [x] `tmp<Field<tensor>> higherOrderGrad::fGradGaussPoints(const volVectorField& D);`
> - [x]  where the returned field has the size `label size = 0; forAll(faceGaussPoints(), faceI){size += faceGaussPoints()[faceI].size();}` with `faceGaussPoints()` being a `List<pointField>` of the Gauss point coordinates per face (or a similar field, if we do not need to store the positions).
> - [x] This `fGradGaussPoints` function will need the coefficients at the face Gauss points. We can create and store these just like we currently do for the cells, however, in this case, we have an extra nested level.

Now, we should proceed with:

> - [x] In the solid model (e.g. `linGeomTotalDispSolid`), we can call the`fGradGaussPoints` function to get gradD at the face Gauss points.
> - [x] We then need to add an interface to the mechanical model (something simple initially for testing) that takes a `Field<tensor>` as an input and returns the stress field as `Field<symmTensor>`, i.e., the stress at the face Gauss points.
> - [ ] Finally, we need to define a function (e.g., `tmp<volVectorField> fvc::divGaussPoints(const Field<vector>& tractionAtGaussPoints)` or similar that lets us integrate the force at the faces. This function will need the Gauss weights, so will have to figure out the best place to calculate and store these (we can pass them as an argument if needed).


Shall we continue communication about this on the Feature petsc SNES higher order pull request?

**Comment by iBatistic** [comment]
​@philipcardiff, can I add commits to the Pull Request that you opened in the solids4foam repository? For example, if I want to commit directly to PR #155.  
If not, the alternative would be to create a PR that updates your PR, is that correct?
